qiskit/pulse/transforms/dag.py

```diff
@@ -64,39 +64,44 @@ def block_to_dag(block: ScheduleBlock) -> rx.PyDAG:
     """
     if block.alignment_context.is_sequential:
         return _sequential_allocation(block)
-    else:
-        return _parallel_allocation(block)
+    return _parallel_allocation(block)
 
 
-def _sequential_allocation(block: ScheduleBlock) -> rx.PyDAG:
+def _sequential_allocation(block) -> rx.PyDAG:
     """A helper function to create a DAG of a sequential alignment context."""
-    dag_blocks = rx.PyDAG()
+    dag = rx.PyDAG()
 
-    prev_node = None
     edges = []
-    for inst in block.blocks:
-        current_node = dag_blocks.add_node(inst)
-        if prev_node is not None:
-            edges.append((prev_node, current_node))
-        prev_node = current_node
-    dag_blocks.add_edges_from_no_data(edges)
+    for elm in block.blocks:
+        node_id = dag.add_node(elm)
+        if dag.num_nodes() == 1:
+            continue
+        prev_id = dag.node_indices()[-1]
+        edges.append((prev_id, node_id))
```

Comment:
I haven't tested it, but this _feels_ very like `prev_id` is just going to be the same as `node_id` every time?  If so, this isn't make a DAG - it's making a graph where every node points only to itself.

Reply:
Hah, you are right. It should be making a DAG that is just a linear chain of all the elements but it is instead making a set of nodes pointing to themselves. I think this indicates a missing test -- if you make two schedule blocks with the same elements but in different orders, a test that checks that the two schedules are not equal should fail.

Reply:
It turns out that there is good test coverage of the different equality cases in test_block.py. Here is the test I was suggesting:

https://github.com/Qiskit/qiskit-terra/blob/fe65c1a2cff8b1061425d92fb03ccc9534c5b1b4/test/python/pulse/test_block.py#L474-L484

However, there is a quirk that makes this test pass. The `if dag.num_nodes() == 1` condition causes the first node not to get an edge to itself. So for two-element schedules, the two elements end up unique -- the first one has no edges while the second one has an edge. This variant of the test fails for me though:

```python
        block1 = pulse.ScheduleBlock(alignment_context=self.sequential_context)
        block1 += pulse.Play(self.test_waveform0, self.d0)
        block1 += pulse.Play(self.test_waveform0, self.d0)
        block1 += pulse.Play(self.test_waveform0, self.d1)

        block2 = pulse.ScheduleBlock(alignment_context=self.sequential_context)
        block1 += pulse.Play(self.test_waveform0, self.d0)
        block2 += pulse.Play(self.test_waveform0, self.d1)
        block2 += pulse.Play(self.test_waveform0, self.d0)

        self.assertNotEqual(block1, block2)
```

In this case, the first element of each schedule is the same so that they compare okay as the no-edge parts of the graphs, and then the remaining elements all point to themselves. `is_isomorphic_node_match` can then reorder the remaining elements to match them since there are no edges between them.

Reply:
In fairness, your second Python block there has a `block1 +=` hidden among the `block2` mutations.  I don't know if that affects anything.

Reply:
Oops, that was me redoing the modification in the comment instead of copy-pasting from the test file I modified. Fixed in the code above now.

Reply:
This can accidentally pass this test. Indeed this always creates closed edges, but this doesn't appear in the first node. So test can still pass up to two nodes. Code is fixed in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/c973d61e2b2fa32cfec94e8340feb9ac2b89613e

Test from @wshanks looks effective to detect the bug. Added in https://github.com/Qiskit/qiskit-terra/pull/8005/commits/1531480175836a568ce1253ce77a328fca0bc983.